File: calculation/RadixSort.cpp
#define MAX_BUCKET 0x100
template<typename T>
void RadixSort(T base[], int n){
	const int max_bit = 8 * sizeof(T);
	const int MASK = MAX_BUCKET-1;

	static T buffer[MAX_LENGTH];
	static T t[MAX_LENGTH];
    static T sa[MAX_LENGTH];	//sa[i] is the index of the i_th smallest number
    static T c[MAX_BUCKET];

	T *a = base, *b = buffer;
	for(int k = 0; k < max_bit; k += 8){
		memset(c, 0, sizeof(c));
		for(int i = 0; i < n; ++i){
			t[i] = (a[i] >> k) & MASK;
			++c[t[i]];
		}
		for(int i = 1; i < MAX_BUCKET; ++i)
			c[i] += c[i-1];
		for(int i = n-1; i >= 0; --i)
			sa[--c[t[i]]] = i;
		for(int i = 0; i < n; ++i)
			b[i] = a[sa[i]];
		swap(a, b);
    }
    if(a != base)
		memcpy(base, a, n * sizeof(T));
}

File: rust/radix_sort.rs
use std::mem;
const MAX_BUCKET: usize = 0x100;
fn radix_sort(base: &mut [usize]) {
    const MAX_BIT: usize = 8 * mem::size_of::<usize>();
    const MASK: usize = MAX_BUCKET - 1;
    let mut a = base;
    let n = a.len();
    let mut buffer = vec![0usize; n];
    let mut b = &mut buffer[..];
    let mut t = vec![0usize; n];
    let mut sa = vec![0usize; n];

    for k in (0..MAX_BIT).step_by(8) {
        let mut c = vec![0usize; MAX_BUCKET];
        for i in 0..n {
            t[i] = (a[i] >> k) & MASK;
            c[t[i]] += 1;
        }
        for i in 1..MAX_BUCKET {
            c[i] += c[i - 1];
        }
        for i in (0..n).rev() {
            c[t[i]] -= 1;
            sa[c[t[i]]] = i;
        }
        for i in 0..n {
            b[i] = a[sa[i]];
        }
        let tmp = a;
        a = b;
        b = tmp;
    }
    if MAX_BIT & 1 != 0 {
        for i in 0..n {
            b[i] = a[i];
        }
    }
}
